### ISSUE TITLE
IDE1006: Allow naming rule exceptions to be configured in .editorconfig

### ISSUE THREAD

**Issue by santo998** [dotnet/roslyn#85090]
**Brief description:**

Allow exceptions to naming rules reported by `IDE1006` to be configured directly in `.editorconfig`.

Currently, naming conventions can be defined using settings such as:

```ini
dotnet_naming_rule.*
dotnet_naming_style.*
dotnet_naming_symbols.*
```

However, there does not appear to be a way to exclude specific symbol names from a naming rule.

A possible configuration could look like:

```ini
dotnet_naming_rule.protected_methods_lowercase.excluded_names = Dispose
```

or, alternatively, something diagnostic-specific such as:

```ini
dotnet_code_quality.IDE1006.excluded_names = Dispose
```

The exact syntax is not important. The main request is to provide a way to define exceptions to naming rules globally in `.editorconfig`.

This would be conceptually similar to analyzer-specific configuration options such as:

```ini
dotnet_code_quality.CA1062.null_check_validation_methods = ...
```

**Languages applicable:**

Both C# and Visual Basic, assuming naming rule configuration is shared between both languages.

My particular use case is C#.

**Code example that the analyzer should report:**

In my projects, `protected` methods intentionally use lowercase names:

```csharp
protected virtual void doSomething()
{
}
```

This convention is enforced through `.editorconfig`.

However, the standard `IDisposable` pattern commonly contains a `protected` method named `Dispose`:

```csharp
public void Dispose()
{
	Dispose(true);
	GC.SuppressFinalize(this);
}

protected virtual void Dispose(bool disposing)
{
	// Cleanup
}
```

Because `Dispose(bool)` is `protected`, my naming rule expects its name to start with a lowercase letter, so `IDE1006` reports it as a naming violation.

Renaming it to:

```csharp
protected virtual void dispose(bool disposing)
{
}
```

would satisfy the configured naming convention, but would deviate from the standard and widely recognized `IDisposable` pattern.

Ideally, I would be able to keep the existing naming rule while adding an exception such as:

```ini
dotnet_naming_rule.protected_methods_lowercase.excluded_names = Dispose
```

Multiple exclusions could also be useful:

```ini
dotnet_naming_rule.protected_methods_lowercase.excluded_names = Dispose, SomeFrameworkMethod
```

**Additional information:**

The current alternatives are to:

- suppress `IDE1006` around individual methods;
- weaken the naming rule;
- disable the rule;
- or add project-specific workarounds.

For naming conventions shared across many projects, these approaches are inconvenient because the exception cannot be centralized in the common `.editorconfig`.

Supporting exclusions would allow custom naming conventions to coexist with framework or library APIs that require, or strongly encourage, specific symbol names.

Although `IDisposable.Dispose(bool)` is my concrete use case, this feature could also be useful for methods, properties, fields, events, or other symbols whose names need to be exempted from an otherwise valid naming rule.

**Documentation requirements:**

When this analyzer is implemented, it must be documented by following the steps at [Documentation for IDE CodeStyle analyzers](https://github.com/dotnet/roslyn/blob/main/docs/contributing/Documentation%20for%20IDE%20CodeStyle%20analyzers.md).

**Comment by MihuBot** [comment]
I'm a bot. Here is a possible related and/or duplicate issue (I may be wrong):
- https://github.com/dotnet/roslyn/issues/46708

**Comment by santo998** [comment]
@MihuBot not really, as this is a FR

**Comment by CyrusNajmabadi** [comment]
> would satisfy the configured naming convention, but would deviate from the standard and widely recognized IDisposable pattern.

I'm not really understanding the issue there. People will certainly understand what's going on. It's not like the casing of the protected method is what's relevant. 

And this is certainly rare enough to be fine to just suppress in place.

**Comment by CyrusNajmabadi** [comment]
I'm not a fan of this as it seems like simple names like this would just be too blunt.  Some cases of 'dispose' should be exempted, while others should not be. 

It's just too coarse grained.

**Comment by santo998** [comment]
> And this is certainly rare enough to be fine to just suppress in place.

In my case, though, that would require a lot of suppressions — one every time I implement `IDisposable`

> I'm not a fan of this as it seems like simple names like this would just be too blunt. Some cases of 'dispose' should be exempted, while others should not be.

`Dispose` is just one example. My proposal is broader: allow exceptions to naming rules to be configured in `.editorconfig`.

**Comment by CyrusNajmabadi** [comment]
> Dispose is just one example. 

I think a lot more examples would be necessary.
